File: egg/egg.py
# https://bernsteinbear.com/blog/whats-in-an-egraph/
from __future__ import annotations
from dataclasses import dataclass
import itertools
from typing import Optional

every_op = []

@dataclass
class Expr:
    forwarded: Optional[Expr] = dataclasses.field(
        default=None,
        init=False,
        compare=False,
        hash=False,
    )

    def __post_init__(self) -> None:
        global every_op
        every_op.append(self)

    def find(self) -> Expr:
        """Return the representative of the set containing `self`."""
        expr = self
        while expr is not None:
            next = expr.forwarded
            if next is None:
                return expr
            expr = next
        return expr

    def make_equal_to(self, other) -> None:
        """Union the set containing `self` with the set containing `other`."""
        found = self.find()
        if found is not other:
            found.forwarded = other

@dataclass
class Const(Expr):
    value: int

@dataclass
class Var(Expr):
    name: str

@dataclass
class BinaryOp(Expr):
    left: Expr
    right: Expr

@dataclass
class Add(BinaryOp):
    pass

@dataclass
class Mul(BinaryOp):
    pass

@dataclass
class Div(BinaryOp):
    pass

@dataclass
class LeftShift(BinaryOp):
    pass


def is_const(op: Expr, value: int) -> bool:
    return isinstance(op, Const) and op.value == value

def optimize_one(op: Expr) -> None:
    if isinstance(op, BinaryOp):
        left = op.left.find()
        right = op.right.find()
        if isinstance(op, Add):
            if isinstance(left, Const) and isinstance(right, Const):
                op.make_equal_to(Const(left.value + right.value))
            elif is_const(left, 0):
                op.make_equal_to(right)
            elif is_const(right, 0):
                op.make_equal_to(left)
        elif isinstance(op, Mul):
            if is_const(left, 1):
                op.make_equal_to(right)
            elif is_const(right, 1):
                op.make_equal_to(left)
            elif is_const(right, 2):
                op.make_equal_to(Add(left, left))
                op.make_equal_to(LeftShift(left, Const(1)))

def optimize(ops: list[Expr]):
    for op in ops:
        optimize_one(op.find())

def eg1():
  ops = [
      a := Const(1),
      b := Const(2),
      c := Add(a, b),
  ]
  print("BEFORE:")
  for op in ops:
      print(f"v{op.id} =", op.find())
  optimize(ops)
  print("AFTER:")
  for op in ops:
      print(f"v{op.id} =", op.find())
eg1()
  

def discover_eclasses(ops: list[Expr]) -> dict[Expr, set[Expr]]:
    eclasses: dict[Expr, set[Expr]] = {}
    for op in ops:
        found = op.find()
        if found not in eclasses:
            # Key by the representative
            eclasses[found] = set()
        eclasses[found].add(op)
        if op is not found:
            # Alias the entries so that looking up non-representatives also
            # finds equivalent operations
            eclasses[op] = eclasses[found]
    return eclasses

def eg2():
  print("ECLASSES:")
  eclasses = discover_eclasses(every_op.copy())
  for op in ops:
      print(f"v{op.id} =", eclasses[op])
eg2()


def eg3():
  ops = [
      a := Var("a"),
      b := Const(2),
      c := Mul(a, b),
      d := Div(c, b),
  ]

def optimize_match(op: Expr, eclasses: dict[Expr, set[Expr]]):
    # Find cases of the form (a * b) / b and rewrite to a
    for e0 in eclasses[op]:
        if isinstance(e0, Div):
            div_left = e0.left
            div_right = e0.right
            for e1 in eclasses[div_left]:
                if isinstance(e1, Mul):
                    mul_left = e1.left
                    mul_right = e1.right
                    if mul_right == div_right:
                        op.make_equal_to(mul_left)
                        return

# https://github.com/bytecodealliance/rfcs/blob/main/accepted/cranelift-egraph.md
def extract(program: list[Expr], eclasses: dict[Expr, set[Expr]]) -> list[Expr]:
    best_cost = float("inf")
    best_program = program
    for trial_program in itertools.product(*[eclasses[op] for op in program]):
        cost = whole_program_cost(trial_program)
        if cost < best_cost:
            best_cost = cost
            best_program = trial_program.copy()
    return best_program

# aegraphs: https://cfallin.org/pubs/egraphs2023_aegraphs_slides.pdf

File: egg/egg.cpp
// https://bernsteinbear.com/blog/whats-in-an-egraph/
#include <iostream>
enum ExprKind {
  None,
  Const,
  Add,
  Mul,
  Div,
  LeftShift,
}
struct Expr {
  Expr *unite; // every expression has a pointer to its "unique representative". (Union-find)
  ExprKind kind;

  Expr(ExprKind kind) {
    this->unite = this;
    this->kind = kind;
  }

  Expr *find() {
    Expr *out = this;
    while true {
      // path compress
      if (out->unite == out) {
        this->unite = out;
        return out;
      } else {
        out = peek;
      }
    }
  }

  Expr *mkEqualTo(Expr *other) {
    Expr *found = this->find();
    assert(found->unite == found);
    if (found != other) {
      found->unite = other;
    }
  }
};

struct ExprConst : public Expr {
  int val;
  ExprConst (int val) : val(val), Expr(ExprKind::Const) {};
}

struct ExprBinop : public Expr {
  Expr *left, *right;
  ExprBinop(ExprKind kind, Expr *left, Expr *right) : Expr(kind, left, right) {}
}

Expr *ExprAdd(Expr *left, Expr *right) {
  return new ExprBinop(ExprKind::Add, left, right);
}

void optimize1(Expr *e) {
  ExprBinop *op = dynamic_cast<ExprBinop>(e);
  if (!op) { return; }
  if (op->kind == ExprKind::Add) {
    ExprConst *l = dynamic_cast<ExprConst>(op->left);
    ExprConst *r = dynamic_cast<ExprConst>(op->left);
    if (l && l->val == 0) {
      op->makeEqualTo(r);
    } else if (r && r->val == 0) {
      op->makeEqualTo(l);
    } else if (l && r) {
      op->makeEqualTo(new Const(l->val + r->val));
    }
  }
}

using CanonCtx = map<Expr *, set<Expr *>*>
CanonCtx discoverEClasses(Expr *op, CanonCtx *ctx) {
  Expr *found = op->find();
  auto it = ctx->find(found);
  if (it == ctx->end()) {
    ctx[found] = new std::set<Expr *>();
    ctx[found]->add(op);
  } else {
    // alias the entries, so that looking up non-representatives also finds equvalent ops.
    ctx[op] = it->second;
  }
}

void optimize2(Expr *e, CanonCtx *ctx) {
  assert(ctx->find(e) != ctx->end());
  // find (a e1bin:* b) e0bin:/ b → a
  for (auto e0 : *ctx->find[e]) {
    BinaryOp *e0bin = dynamic_cast<BinaryOp>(e0);
    if (!e0bin) { continue; }
    if (e0bin->kind != ExprKind::Div) { continue; }
    // lookup the left argument, see if it has a (a * b) inside it.
    // But this lookup is slow!
    // I am pretty sure that this is the part that is hard to get right,
    // and that the 'egg' folks have a different way of doing the search?
    // or am I crazy?
    // I guess I can store the nodes in a trie :)
    for (auto e1 : *ctx->find[e0bin->left]) {
      BinaryOp *e1bin = dynamic_cast<BinaryOp>(e1);
      if (!e1bin) { continue; }
      if (e1bin->kind != ExprKind::Div) { continue; }
      if (e1bin->rhs->find() != e0bin->rhs->find()) { continue: }
      e->mkEqualTo(e1bin->lhs);
    }
  }
}

// (?a + 0) -> ?a
// subtitution map.
struct Matcher;
struct Subst {
  map<Matcher *, Expr *> ctx;
  bool add(Matcher *m, Expr *e) {
    auto it = ctx->find(m);
    if (it == m->end()) {
      ctx[m] = e;
      return true;
    }
    // found an existing match for 'm', check that it's the same expression.
    if (it->second == e) { return true; }
    // mismatch, return false.
    return false;
  }

  optional<Expr *> find(Matcher *m) {
    auto it = ctx->find(m);
    if (it == m->end()) {
      return {};
    } else {
      return {it->second;}
    }
  }
}

// "Term index"
struct Matcher {
  // match an expression, updating the substitution map,
  // and returning 'false' if no substitution was possible.
  virtual bool match(Expr *e, Subst &s) = 0; 
};

struct AnyMatcher : public Matcher {
  bool match(Expr *e, Subst &s) {
    s.add(this, e);
  }
}


struct ConstMatcher : public Matcher {
  optional<int> v;
  ConstMatcher() {}
  ConstMatcher(optional<int> v) : v(v) {}

  bool match(Expr *e, Subst &s) {
    ExprConst *c = dynamic_cast<ExprConst>(e);
    if (!c) { return false; }
    // this->v => c->v == this->v
    if (!this->v || c->v != *this->v) { return false; }
    s.add(this, c);
    return true;
  }
}

struct BinopMatcher : public Matcher {
  Matcher *left, *right;
  BinopMatcher(Matcher *left, Matcher *right) : left(left), right(right) {}
};

